### PR TITLE
Fix merged line comments

### LINKED ISSUE #9959: Single line comments are merged on trailing trivia
**Prettier 2.1.2**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAzArlMMCW0AEAHgBQCU+wAOlAL74D09+qOATgM4zUDcj+7cSFAAmIADQgIAB1zR2yUAENWrCAHcACsoTyUigDZrFAT3kSARq0VgA1nBgBlKdZxQA5shit0cCXAC25nDCwsEAMoru6IpucABiEKz+ijC47sggiugwEOIgABYw-voA6vk48OzOYHAOOhU4AG4Vxhlg7GYgrgKsMBpWbsnIqAYCEgBW7IQAQla29g6K-nBhrnDDo74gk4QOrm76cACK6BDwG-pjIM4ccKwZ5opB+nlSrK4wJTjCMPnIABwABgkbwgAhKVikGTecB6jXWEgAjqd4P1pLpMuwALRQODBYJ5VhwZFsOD9GJDJAjS5bAT+HCeby0-aHE5ndZUzYSGBPL4-P5IABM3KsOH0+wAwhB-JSQLCAKx5dACAAqT101KujR8AEkRAhHGB3jIAIIiBwwYyHC4CGg0IA)
```sh
--parser babel
```

**Input:**
```jsx
function x() {
} // first
; // second
```

**Output:**
```jsx
function x() {} // first // second

```

**Expected behavior:**
```jsx
function x() {} // first
// second
```

**Details:**

As you can see, the two comments (that should be on their own lines) are merged on a single line. I think this has to do with the `lineSuffix` command being used on two tokens on the same line. When that happens, both the suffixes are appended to the same line.

This is obviously very dangerous, and destroys the semantic meaning of the comment. Not only it makes it much harder to read and reason about (since comments cannot be broken), it also means that it will break any tools/parsers that read the comment contents and expect certain strings to be there/not there, like ignores and other directives.

| - | First | Second |
| --------------- | --------------- | --------------- |
| Before | ` first` | ` second` |
| After | ` first // second` | **removed** |

This does not just affect JavaScript, but it also affects all other languages, as they have no other means AFAICT of fixing this correctly using other available commands.

### PR DESCRIPTION
**Pull request by thorn0**
## Description

Fixes #9959

## Checklist

- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: src/main/comments.js [hunk 1]
```diff
@@ -461,14 +463,30 @@ function printLeadingComment(path, options) {
...
 
-  if (hasNewline(originalText, locStart(comment), { backwards: true })) {
+  let previousCommentHasLineSuffix = false;
+
+  if (previousComment) {
+    const previousCommentDocType = getDocType(previousComment.doc);
+    if (
+      previousCommentDocType === DOC_TYPE_LINE_SUFFIX ||
+      (previousCommentDocType === DOC_TYPE_ARRAY &&
+        getDocType(previousComment.doc[0]) === DOC_TYPE_LINE_SUFFIX)
```
Comment (fisker) [inline src/main/comments.js:479]: Should we use `findInDoc`? The printed doc may change in future.

**Comment by fisker** [comment]
Since you already got previous comment, may a simple check here https://github.com/thorn0/prettier/blob/ed63ad5928f5ba1dd60e5e6c0fda5906584b3989/src/main/comments.js#L511 will fix #12653?

**Comment by thorn0** [comment]
> Since you already got previous comment, may a simple check here https://github.com/thorn0/prettier/blob/ed63ad5928f5ba1dd60e5e6c0fda5906584b3989/src/main/comments.js#L511 will fix #12653?

#12653 is about leading comments too, not only trailing ones. Let's fix it separately. BTW, we're going to need a new printer method for that check because it's a JS-specific thing.

File: src/main/comments.js (hunk 1 above)
Comment (thorn0) [reply src/main/comments.js:479]: The tests should save us. `findInDoc` is recursive. Seems like too much overhead.

As an alternative solution, what if we change the return type of this function to `{ doc, isBlock, hasLineSuffix }`

**Review by fisker — APPROVED** [review]

File: src/main/comments.js (hunk 1 above)
Comment (fisker) [reply src/main/comments.js]: Much better!

**Pull request merged**